tests/kafkatest/tests/client/downconversion_memory_test.py

```diff
@@ -0,0 +1,119 @@
+    def __init__(self, test_context):
+        super(DownconversionMemoryTest, self).__init__(test_context=test_context)
+        '''
+        Test Setup:
+        ==========
+        - Java heap size = 190MB
+        - 1M messages, 1kB each ==> 1GB of total messages
+        - Split into 200 partitions ==> approximately 5MB per partition
+        - 1 consumer with `fetch.max.bytes` = 200MB and `max.partition.fetch.bytes` = 1MB
+        - Each fetch consumes min(1MB*200, 200MB) = 200MB i.e. 1MB from each partition for a total of 200MB
+        - Success criteria:
+            - Must always run out of memory if not using lazy down-conversion
+            - Must never run out of memory if using lazy down-conversion
+        '''
+        self.heap_size = 190
+        self.max_messages = 1024 * 1024
+        self.message_size = 1024
+        self.batch_size = self.message_size * 50
+        self.num_partitions = 200
+        self.max_fetch_size = 200 * 1024 * 1024
+        self.num_producers = 1
+        self.num_consumers = 1
+        self.topics = ["test_topic"]
+        self.zk = ZookeeperService(self.test_context, num_nodes=1)
+        self.heap_dump_path = "/mnt/heapdump/"
+        self.consumer_version = LATEST_0_10
+
+    def setUp(self):
+        self.zk.start()
+        super(DownconversionMemoryTest, self).setUp()
+
+    def tearDown(self):
+        for node in self.kafka.nodes:
+            node.account.ssh("rm -rf %s" % self.heap_dump_path, allow_fail=False)
+        super(DownconversionMemoryTest, self).tearDown()
+
+    @cluster(num_nodes=12)
+    @parametrize(version=str(LATEST_1_1))
+    @parametrize(version=str(DEV_BRANCH))
+    def test_downconversion(self, version):
+        broker_version = KafkaVersion(version)
+        producer_version = broker_version
+
+        out_of_memory = False
+        if KafkaVersion(version) <= LATEST_1_1:
+            expect_out_of_memory = True
+            consumer_timeout_sec = 60
+        else:
+            expect_out_of_memory = False
+            consumer_timeout_sec = None
+
+        self.kafka = KafkaService(self.test_context, num_nodes=1, zk=self.zk, version=broker_version,
+                                  topics={topic:
+                                              {"partitions": self.num_partitions,
+                                               "replication-factor": 1,
+                                               "configs": {"min.insync.replicas": 1}}
+                                          for topic in self.topics},
+                                  heap_opts="-Xmx%dM -Xms%dM -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=%s" % (self.heap_size, self.heap_size, self.heap_dump_path),
+                                  do_logging=False)
+        for node in self.kafka.nodes:
+            node.account.ssh("mkdir -p %s" % self.heap_dump_path, allow_fail=False)
+        self.kafka.start()
+
+        # seed kafka with messages
+        for topic in self.topics:
```

Comment:
nit: these tests are easier to understand if we try to break them into smaller pieces. For example, we could turn this into a separate function.